central/group/datastore/datastore_impl.go

```diff
@@ -70,18 +101,26 @@ func (ds *dataStoreImpl) Add(ctx context.Context, group *storage.Group) error {
 		return errox.InvalidArgs.CausedBy(err)
 	}
 
-	if group.GetProps().GetId() != "" {
-		return errox.InvalidArgs.Newf("id should be empty but %q was provided", group.GetProps().GetId())
+	if err := setGroupIDForNewGroup(group); err != nil {
+		return err
 	}
 
-	if group.GetProps() != nil {
-		group.GetProps().Id = GenerateGroupID()
-	} else {
-		// Theoretically should never happen, as the auth provider ID is required to be set.
-		group.Props = &storage.GroupProperties{Id: GenerateGroupID()}
+	// Lock to ensure synchronization between the get and upsert
```

Comment:
The comment indicates that the lock is to sync gets and upserts but the getters do not hold any lock. Nonetheless, sync between get and upsert is very strict. I recommend not locking in getters.

Reply:
By "get" I mean the fact that this add is getting something (Default groups - see line 112), **not getters**

Reply:
I reworded the comment